macrosynergy/download/dq_api.py

```diff
@@ -0,0 +1,629 @@
+    """Will return the request from DataQuery."""
+    request_error = f"Unknown request method {method} not in ('get', 'post')."
+    assert method in ("get", "post"), request_error
+
+    with requests.request(
+        method=method,
+        url=url,
+        cert=cert,
+        headers=headers,
+        params=params,
+        **kwargs,
+    ) as r:
+        text: str = r.text
+        last_url: str = r.url
+        js, success, msg = valid_response(r=r)
+
+    return js, success, last_url, msg
+
+
+class CertAuth(object):
+    """Certificate Authentication.
+
+    Class used to access DataQuery via certificate and private key. To access the API
+    login both username & password are required as well as a certified certificate and
+    private key to verify the request.
+
+    :param <str> username: username for login to REST API for JP Morgan DataQuery.
+    :param <str> password: password.
+    :param <str> crt: string with location of public certificate.
+    :param <str> key: string with private key location.
+
+    """
+
+    def __init__(
+        self,
+        username: str,
+        password: str,
+        crt: str = "api_macrosynergy_com.crt",
+        key: str = "api_macrosynergy_com.key",
+        base_url: str = CERT_BASE_URL,
+    ):
+
+        error_user = f"username must be a <str> and not <{type(username)}>."
+        assert isinstance(username, str), error_user
+
+        error_password = f"password must be a <str> and not <{type(password)}>."
+        assert isinstance(password, str), error_password
+
+        self.auth: str = base64.b64encode(
+            bytes(f"{username:s}:{password:s}", "utf-8")
+        ).decode("ascii")
+
+        self.headers: Dict[str, str] = {"Authorization": f"Basic {self.auth:s}"}
+        self.base_url: str = base_url
+
+        # Key and Certificate.
+        self.key: str = self.valid_path(key, "key")
+        self.crt: str = self.valid_path(crt, "crt")
+
+        # For debugging purposes save last request response.
+        self.status_code: Optional[int] = None
+        self.last_response: Optional[str] = None
+        self.last_url: Optional[str] = None
+
+    @staticmethod
+    def valid_path(directory: str, file_type: str) -> Optional[str]:
+        """Validates the key & certificate exist in the referenced directory.
+
+        :param <str> directory: directory hosting the respective files.
+        :param <str> file_type: parameter used to distinguish between the certificate or
+            key being received.
+
+        """
+        dir_error = f"{file_type:s} file must be a <str> not <{type(directory)}>."
+        assert isinstance(directory, str), dir_error
+
+        condition = os.path.exists(directory) and os.path.isfile(directory)
+        if not condition:
+            raise OSError(
+                f"The directory received, {directory}, does not contain the "
+                f"respective file, {file_type}."
+            )
+
+        return directory
```

Comment:
I think we need to be clearer in our language, as isn't technically what we are checking here whether the filename (and path of the file) is valid and exists.